File: packages/android/src/main/java/com/kozmos/components/Tag/Tag.kt
package com.kozmos.components.tag

import com.kozmos.tokens.KozmosDimensions

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.kozmos.components.KozmosEmotion
import com.kozmos.tokens.KozmosColors

enum class KozmosTagVariant {
    Default,
    Secondary,
    Destructive,
    Outline
}

@Composable
fun KozmosTag(
    text: String,
    variant: KozmosTagVariant = KozmosTagVariant.Default,
    /**
     * Unset and the variant draws as it always has; set and the emotion
     * decides the colour, whatever the variant.
     */
    emotion: KozmosEmotion? = null,
    onRemove: (() -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    val shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusContainer)
    val emotionSurface = emotion?.surface
    val emotionOnSurface = emotion?.onSurface
    val emotionText = emotion?.text
    val backgroundColor = when (variant) {
        KozmosTagVariant.Default -> KozmosColors.primitivesColorsTheme500
        KozmosTagVariant.Secondary -> KozmosColors.primitivesColorsBackground100
        KozmosTagVariant.Destructive -> KozmosColors.semanticsDataRed
        KozmosTagVariant.Outline -> Color.Transparent
    }
    val foregroundColor = when (variant) {
        KozmosTagVariant.Default,
        KozmosTagVariant.Destructive -> KozmosColors.primitivesColorsBackground0
        KozmosTagVariant.Secondary,
        KozmosTagVariant.Outline -> KozmosColors.primitivesColorsForeground900
    }
    val resolvedBackground = when {
        emotion == null -> backgroundColor
        variant == KozmosTagVariant.Outline -> Color.Transparent
        else -> emotionSurface!!
    }
    val resolvedForeground = when {
        emotion == null -> foregroundColor
        variant == KozmosTagVariant.Outline -> emotionText!!
        else -> emotionOnSurface!!
    }
    val borderedModifier = if (variant == KozmosTagVariant.Outline) {
        Modifier.border(
            width = 1.dp,
            color = emotionText ?: KozmosColors.primitivesColorsBackground200,
            shape = shape
        )
    } else {
        Modifier
    }

    Row(
        modifier = modifier
            .background(resolvedBackground, shape)
            .then(borderedModifier)
            .padding(horizontal = KozmosDimensions.primitivesLayoutSpacing100, vertical = KozmosDimensions.primitivesLayoutSpacing50),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = text,
            color = resolvedForeground,
            fontSize = 12.sp,
            fontWeight = FontWeight.SemiBold
        )
        if (onRemove != null) {
            Icon(
                imageVector = Icons.Default.Close,
                contentDescription = "Remove",
                tint = foregroundColor,
                modifier = Modifier
                    .padding(start = KozmosDimensions.primitivesLayoutSpacing50)
                    .size(KozmosDimensions.primitivesLayoutSizing200)
                    .clip(CircleShape)
                    .clickable { onRemove() }
            )
        }
    }
}

File: packages/ios/Sources/Components/Tag/Tag.swift
import SwiftUI

/// What a tag means, as opposed to how it is drawn.
///
/// The product drives this axis on 33,988 tag instances across ten surfaces.
/// The colours come from `Semantics.Emotion`, which exists so a tag does not
/// have to borrow a button's tokens.
public enum KozmosEmotion {
    case neutral, themed, success, danger, informative, alert

    var surface: Color {
        switch self {
        case .neutral: return KozmosColors.semanticsEmotionNeutralSurface
        case .themed: return KozmosColors.semanticsEmotionThemedSurface
        case .success: return KozmosColors.semanticsEmotionSuccessSurface
        case .danger: return KozmosColors.semanticsEmotionDangerSurface
        case .informative: return KozmosColors.semanticsEmotionInformativeSurface
        case .alert: return KozmosColors.semanticsEmotionAlertSurface
        }
    }

    var onSurface: Color {
        switch self {
        case .neutral: return KozmosColors.semanticsEmotionNeutralOnsurface
        case .themed: return KozmosColors.semanticsEmotionThemedOnsurface
        case .success: return KozmosColors.semanticsEmotionSuccessOnsurface
        case .danger: return KozmosColors.semanticsEmotionDangerOnsurface
        case .informative: return KozmosColors.semanticsEmotionInformativeOnsurface
        case .alert: return KozmosColors.semanticsEmotionAlertOnsurface
        }
    }

    /// The emotion on the page itself, at the first step of each ramp that
    /// reaches 4.5:1 — which is not the same step for every emotion.
    var text: Color {
        switch self {
        case .neutral: return KozmosColors.semanticsEmotionNeutralText
        case .themed: return KozmosColors.semanticsEmotionThemedText
        case .success: return KozmosColors.semanticsEmotionSuccessText
        case .danger: return KozmosColors.semanticsEmotionDangerText
        case .informative: return KozmosColors.semanticsEmotionInformativeText
        case .alert: return KozmosColors.semanticsEmotionAlertText
        }
    }
}

public struct KozmosTag: View {
    let text: String
    let variant: KozmosTagVariant
    /// Unset and the variant draws as it always has; set and the emotion
    /// decides the colour, whatever the variant.
    let emotion: KozmosEmotion?
    let onRemove: (() -> Void)?
    
    public enum KozmosTagVariant {
        case `default`, secondary, destructive, outline, success
        
        var backgroundColor: Color {
            switch self {
            case .default: return KozmosColors.primitivesColorsTheme500
            case .secondary: return KozmosColors.primitivesColorsBackground100
            case .destructive: return KozmosColors.semanticsDataRed
            case .outline: return Color.clear
            case .success: return KozmosColors.semanticsDataTeal
            }
        }
        
        var foregroundColor: Color {
            switch self {
            case .default, .destructive, .success: return KozmosColors.primitivesColorsBackground0
            case .secondary, .outline: return KozmosColors.primitivesColorsForeground900
            }
        }
    }
    
    public init(
        _ text: String,
        variant: KozmosTagVariant = .default,
        emotion: KozmosEmotion? = nil,
        onRemove: (() -> Void)? = nil
    ) {
        self.text = text
        self.variant = variant
        self.emotion = emotion
        self.onRemove = onRemove
    }

    private var resolvedBackground: Color {
        guard let emotion else { return variant.backgroundColor }
        return variant == .outline ? Color.clear : emotion.surface
    }

    private var resolvedForeground: Color {
        guard let emotion else { return variant.foregroundColor }
        return variant == .outline ? emotion.text : emotion.onSurface
    }
    
    public var body: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing50) {
            Text(text)
                .font(KozmosTypography.caption)
                .bold()
                .foregroundColor(resolvedForeground)
            
            if let onRemove = onRemove {
                Button(action: onRemove) {
                    Image(systemName: "xmark")
                        .font(KozmosTypography.caption2)
                        .fontWeight(.bold)
                        .foregroundColor(resolvedForeground)
                }
            }
        }
        .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing100)
        .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing50)
        .background(resolvedBackground)
        .cornerRadius(KozmosDimensions.primitivesLayoutSpacing150)
        .overlay(
            RoundedRectangle(cornerRadius: KozmosDimensions.primitivesLayoutSpacing150)
                .stroke(variant == .outline ? resolvedForeground.opacity(emotion == nil ? 0.3 : 1) : Color.clear, lineWidth: variant == .outline ? 1 : 0)
        )
    }
}
